### PR TITLE
Fix misdetection of some 3.6 typing[_extensions]

### LINKED ISSUE #403: pydantic BaseModel cannot be pickled w/ cloudpickle >= 1.3
## Example

```python
import cloudpickle
import pydantic

class Bar(pydantic.BaseModel):
    pass

if __name__ == '__main__':
    cloudpickle.loads(cloudpickle.dumps(Bar()))
```

## Technical Background
The issue is the interaction between:

https://github.com/samuelcolvin/pydantic/blob/de0657e4a5495de9378db9fe15a17334c2b0fae5/pydantic/main.py#L865-L868

and:

https://github.com/cloudpipe/cloudpickle/blob/cdc704d8df18306f40a80cf7883e000219868dc1/cloudpickle/cloudpickle.py#L460

## Refs
This was originally discussed in #393.

### PR DESCRIPTION
**Pull request by marco-neumann-by**
Fixes #403

### CODE REVIEW

**Comment by marco-neumann-by** [comment]
I honestly don't really understand the issue with the Python 3.5. I also cannot reproduce it locally :(

**Comment by pierreglaser** [comment]
> I honestly don't really understand the issue with the Python 3.5. I also cannot reproduce it locally :(

It seems that the utility `_lookup_module_and_qualname` fails to lookup the module of `typing.AnyString` in `python3.5`. 
It seems to be a `macos`-only issue. I'll see if I can reproduce on my apple machine.

**Comment by pierreglaser** [comment]
@marco-neumann-by do you mind explaining the link between these changes and the original #403 issue? I don't know `pydantic`, so that's not obvious to me.

**Comment by pierreglaser** [comment]
> It seems that the utility _lookup_module_and_qualname fails to lookup the module of typing.AnyString in python3.5.

Digging further into this: the CI failure is a non-deterministic bug caused by our port of `_whichmodule`. I'll open a separate PR for this. Along with a CI fix for `python-nightly` in present in #411 which I will port too, this will make the CI pass on `cloudpickle` :) `master` :)

**Comment by marco-neumann-by** [comment]
@pierreglaser the original issue has a section "Technical Background". Does this already help (it links the relevant pydantic code as well) or shall I go more into detail here?

**Comment by pierreglaser** [comment]
> @pierreglaser the original issue has a section "Technical Background". Does this already help (it links the relevant pydantic code as well) or shall I go more into detail here?

Ah, great, thanks a lot. That makes sense now. Feel free to rebase on #412 to check the CI status on your PR.

File: cloudpickle/cloudpickle.py [hunk 1]
```diff
@@ -455,15 +455,28 @@ def _extract_class_dict(cls):
 
 if sys.version_info[:2] < (3, 7):  # pragma: no branch
     def _is_parametrized_type_hint(obj):
-        # This is very cheap but might generate false positives.
+        # This is very cheap but might generate false positives. So try to narrow it down is good as possible
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py:458]:
```suggestion
        # This is very cheap but might generate false positives. So try to
        # narrow it down is good as possible.
```

File: cloudpickle/cloudpickle.py [hunk 2]
```diff
@@ -455,15 +455,28 @@ def _extract_class_dict(cls):
...
 
         # typing_extensions.Final
-        is_final = getattr(obj, '__type__', None) is not None
+        is_final = (
+            (getattr(obj, '__type__', None) is not None)
+            and from_typing_extensions
+        )
+
+        # typing.ClassVar
+        is_classvar = (getattr(obj, '__type__', None) is not None) and from_typing
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py:479]:
```suggestion
        is_classvar = (
            (getattr(obj, '__type__', None) is not None) and from_typing
        )
```

**Review by ogrisel — APPROVED** [review]
LGTM. I will push a quick changelog entry before merging.

**Pull request merged**